File: neetcode-road-map/two-pointers/trapping-rain-water.cpp
#include <stdlib.h>
#include <stdio.h>
#include <vector>
#include <iostream>
class Solution1 {
public:
    int trap(std::vector<int>& height) {
        int max_l[height.size()];
        int max_r[height.size()];

        max_l[0] = 0; 
        for (size_t i = 1; i < height.size(); i++) {
            max_l[i] = std::max(height[i - 1],max_l[i - 1]);
        }
        max_r[height.size() - 1] = 0;
        for (int i = height.size() - 2; i >= 0; i--) {
            max_r[i] = std::max(height[i + 1],max_r[i + 1]);
        }
        
        int traped = 0;
        for (size_t i = 0; i < height.size(); i++) { 
            max_r[i] = std::min(max_r[i],max_l[i]);
            traped += std::max(0,max_r[i] - height[i]);
        }
        return traped;
    }
};
class Solution {
public:
    int trap(std::vector<int>& height) {
        int traped = 0;
        int last_l = height[0],
            last_r = height[height.size() - 1];
        int l  = 0,
            r  = height.size() - 1;

        while(r > l) {
            if(last_l < last_r) {
                l++;
                last_l  = std::max(last_l,height[l]);
                traped += last_l - height[l];
            } else {
                r--;
                last_r  = std::max(last_r,height[r]);
                traped += last_r - height[r];
            }
        }
        return traped;
    }
};


int main() {
    std::vector<int> height = {0,1,0,2,1,0,1,3,2,1,2,1};
    std::cout << Solution().trap(height)   << "\n";
    return 0;
}

File: neetcode-road-map/two-pointers/trapping-rain-water.py

class Solution:
    def trap(self, height: list[int]) -> int:
        if len(height) <= 2: return 0 

        traped = 0        
        a , b  = 0, 0
        size = len(height)

        while a < size :
            if(height[a] == 0): 
                a +=1
                continue

            k = height[a]
            b = a + 1

            # stop when u find bigger
            # or stop when u out of range
            while(b < size and height[a] > height[b]): 
                k += height[b]
                b += 1


            if(b < size):
                # u found bigger
                traped += (b - a - 1) * min(height[a],height[b]) - k
                a = b
                continue
            else:
                # u didnt find bigger
                # check out the max and calculate using it
                # 4 1 3
                c  = max(height[a:b-1])
                d = min(height[a:b-1])
                if(height[b - 1] != 0):
                    traped += (b - a - 1) * c - k
                

            a += 1
        return traped
height = [0,1,0,2,1,0,1,3,2,1,2,1]
print(Solution().trap(height))